Craig wampler sentinel one2 1 0

CraigWampler opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
None

## Description
Integrations
SentinelOne v2
Added the following commands:
sentinelone-get-blocklist - List out contents of a block list
sentinelone-add-hash-to-blocklist - Add a hash to a block list
sentinelone-remove-hash-from-blocklist - Remove a hash from a block list
sentinelone-fetch-file - Initiate a file fetch command against an endpoint
sentinelone-download-fetched-file - Download a file fetched with the sentinelone-fetch-file command.
Note: SentinelOne uses the term "blacklist" in their documentation/API. This was modified to use the term "blocklist" in the integration commands.

- Updated the Docker image to: demisto/python3:3.10.1.26972.
- Added three playbooks to show how to use the sentinelone-fetch-file and sentinelone-download-fetched-file commands
## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [x] 6.5.0 (only tested with this version)

## Does it break backward compatibility?
   - [x] Maybe? 
       -  Further details:
            I noticed that for the SentinelOne Integration, the name field is SentinelOne V2, but the display is SentinelOne v2. This change in upper/lower casing made it unintuitive to filter to see if the integration was enabled using modules.brand. I changed the case to match (went with lower case). Not sure if that will break code for someone who assumed it would be uppercase or not.
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ x] Tests
The only additional "test" I created are the three attached playbook that validate the file download capability           
- [ x ] Documentation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=18609) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=18609) before we can accept your contribution.<br/><hr/>**Craig Wampler** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=18609) it.</sub>

MLainer1 reviewed (changes requested):
Good work, make sure to look at my suggestions.
Please add a test for each new command, and modify the pack-metadata for the new version.
Also make sure to add the new playbooks to the release notes and look at the checks fails in the pr and fix them.

CraigWampler commented on Packs/SentinelOne/Integrations/SentinelOne-V2/SentinelOne-V2.py:
```diff
@@ -1241,6 +1322,211 @@ def get_processes(client: Client, args: dict) -> CommandResults:
+    sort_by = "updatedAt"
+    sort_order = "desc"
+
+    offset = int(args.get('offset', "0"))
+    limit = int(args.get('limit', "100"))
```
Will change

CraigWampler commented on Packs/SentinelOne/Integrations/SentinelOne-V2/SentinelOne-V2.yml:
```diff
@@ -65,7 +65,7 @@ configuration:
 description: Use the SentinelOne integration to send requests to your management server and get responses with data pulled from agents or from the management database.
 display: SentinelOne v2
-name: SentinelOne V2
+name: SentinelOne v2
```
Will change

CraigWampler commented on Packs/SentinelOne/Integrations/SentinelOne-V2/SentinelOne-V2.yml:
```diff
@@ -1,6 +1,6 @@
 category: Endpoint
 commonfields:
-  id: SentinelOne V2
+  id: SentinelOne v2
```
will change

CraigWampler commented on Packs/SentinelOne/Integrations/SentinelOne-V2/SentinelOne-V2.py:
```diff
@@ -1345,6 +1631,11 @@ def main():
             'sentinelone-threat-summary': get_threat_summary_command,
+            'sentinelone-add-hash-to-blocklist': add_hash_to_blocklist,
+            'sentinelone-remove-hash-from-blocklist': remove_hash_from_blocklist,
+            'sentinelone-get-blocklist': get_blocklist,
```
@MLainer1  all three of those commands are in the YAML file:

sentinelone-add-hash-to-blocklist: line 1547
sentinelone-remove-hash-from-blocklist: line 1564
sentinelone-get-blocklist: Line 1529

Is there another function you were referring to?

CraigWampler commented:
> 

Regarding the failing checks:

Unit tests are failing (before I added any tests) due to the `register_module_line('SentinelOne V2', 'start', __line__())` line. I'm assuming this is an issue for all integrations? How do you normally work around this for testing?

Validation is failing because of many different YAML validation errors. The YAML itself was generated by XSOAR.... so I'm assuming those are okay to ignore?

MLainer1 commented:
> > 
> 
> Regarding the failing checks:
> 
> Unit tests are failing (before I added any tests) due to the `register_module_line('SentinelOne V2', 'start', __line__())` line. I'm assuming this is an issue for all integrations? How do you normally work around this for testing?
> 
> Validation is failing because of many different YAML validation errors. The YAML itself was generated by XSOAR.... so I'm assuming those are okay to ignore?

Make sure to address all of the validation errors, and I'll check about the line error

CraigWampler commented:
> > > 
> > 
> > 
> > Regarding the failing checks:
> > Unit tests are failing (before I added any tests) due to the `register_module_line('SentinelOne V2', 'start', __line__())` line. I'm assuming this is an issue for all integrations? How do you normally work around this for testing?
> > Validation is failing because of many different YAML validation errors. The YAML itself was generated by XSOAR.... so I'm assuming those are okay to ignore?
> 
> @CraigWampler Make sure to address all of the validation errors, and I'll check about the line error. Make sure to merge from master

I fixed what I think are all the issues I can fix.

I also merged in master, but now it's saying that 253 files were changed, so I'm thinking maybe I did the merge wrong? Here's what I did:

```
<pulled lastest master into CraigWampler:master>
git checkout master
git pull
git checkout CraigWampler_SentinelOne2_1_0
git merge master
git push

```

Was I supposed to do something different?

MLainer1 commented:
> Details

You can ignore that :)

MLainer1 commented:
> > > > 
> > > 
> > > 
> > > Regarding the failing checks:
> > > Unit tests are failing (before I added any tests) due to the `register_module_line('SentinelOne V2', 'start', __line__())` line. I'm assuming this is an issue for all integrations? How do you normally work around this for testing?
> > > Validation is failing because of many different YAML validation errors. The YAML itself was generated by XSOAR.... so I'm assuming those are okay to ignore?
> > 
> > 
> > @CraigWampler Make sure to address all of the validation errors, and I'll check about the line error. Make sure to merge from master
> 
> I fixed what I think are all the issues I can fix.
> 
> I also merged in master, but now it's saying that 253 files were changed, so I'm thinking maybe I did the merge wrong? Here's what I did:
> 
> ```
> <pulled lastest master into CraigWampler:master>
> git checkout master
> git pull
> git checkout CraigWampler_SentinelOne2_1_0
> git merge master
> git push
> ```
> 
> Was I supposed to do something different?

You can ignore that :), make sure to look at the Flake8 errors and the validation fails, you need to fix them before the merging

MLainer1 commented on Packs/SentinelOne/Integrations/SentinelOne-V2/SentinelOne-V2.py:
```diff
@@ -1380,3 +1671,5 @@ def main():
 
 if __name__ in ['__main__', 'builtin', 'builtins']:
     main()
+
+# register_module_line('SentinelOne V2', 'end', __line__())
```
```suggestion
```
Please remove these lines

CraigWampler replied:
@MLainer1  - I accidentally left in my comments to remove the `register_module_line` function calls - which relates to your suggested change. In general, my understanding is that they should be there, but that caused pytest to fail. So confirming you want them completely removed?

Also, the remaining validations that are failing are due to YAML issues in playbooks. The YAML was generated by XSOAR. It seems more likely that I would introduce errors if I try to manually resolve each of those. But if you are saying I should try to add fields that it is complaining about, I will take a stab at it. Let me know.

MLainer1 replied:
@[CraigWampler](https://github.com/CraigWampler) I'll look into the yml validations failure, but they will still have to be resolved to be able to merge your pr, so it maybe necessary to generate the yml file again (or add the fields  manually)

MLainer1 commented on Packs/SentinelOne/ReleaseNotes/2_1_0.md:
```diff
@@ -0,0 +1,12 @@
+##### SentinelOne v2
```
Please regenerate the release notes file to match the xsoar standard, using the update-release-notes command

MLainer1 commented on Packs/SentinelOne/Playbooks/playbook-SentinelOne_-_Fetch_File_from_Endpoint.yml:
```diff
@@ -0,0 +1,828 @@
+id: 0d00527a-af88-408d-8c8e-6950dc580493
```
Try to run the `format` command on this file to fix the failed validations

MLainer1 commented on Packs/SentinelOne/Playbooks/playbook-SentinelOne_-_Check_for_New_File_Upload_Activity_README.md:
```diff
@@ -0,0 +1,43 @@
+Manages a polling loop to check for the results of a submitted file fetch request to an endpoint by SentinelOne.
```
Try to run the `format` command on this file to fix the failed validations

MLainer1 commented on Packs/SentinelOne/Playbooks/playbook-SentinelOne_-_Check_for_New_File_Upload_Activity.yml:
```diff
@@ -0,0 +1,280 @@
+contentitemexportablefields:
```
Try to run the `format` command on this file to fix the failed validations

CraigWampler commented:
@tomer-pan: @MLainer1  said to ping you about the CircleCI failures related to playbooks:

https://app.circleci.com/pipelines/github/demisto/content/162302/workflows/50568f28-0151-4b88-9f61-039fb4ea2c50/jobs/515945?invite=true#step-108-73 - These commands are in the readme, and the ReadMe was generated using `update-release-notes`, so I'm not sure why they are not found

https://app.circleci.com/pipelines/github/demisto/content/162302/workflows/50568f28-0151-4b88-9f61-039fb4ea2c50/jobs/515945?invite=true#step-108-80 - Not sure what issue is?

https://app.circleci.com/pipelines/github/demisto/content/162302/workflows/50568f28-0151-4b88-9f61-039fb4ea2c50/jobs/515945?invite=true#step-108-94, 
https://app.circleci.com/pipelines/github/demisto/content/162302/workflows/50568f28-0151-4b88-9f61-039fb4ea2c50/jobs/515945?invite=true#step-108-102,
https://app.circleci.com/pipelines/github/demisto/content/162302/workflows/50568f28-0151-4b88-9f61-039fb4ea2c50/jobs/515945?invite=true#step-108-114 --> DeleteContext. These are all called in sub-playbooks, and `auto` is set to True, so this is deleting the sub-playbook context, which is the goal (e.g if this was called in a loop). So this behavior is by design, not an error.


How do I move ahead with resolving these?

MLainer1 commented:
> @tomer-pan: @MLainer1 said to ping you about the CircleCI failures related to playbooks:
> 
> https://app.circleci.com/pipelines/github/demisto/content/162302/workflows/50568f28-0151-4b88-9f61-039fb4ea2c50/jobs/515945?invite=true#step-108-73 - These commands are in the [ReadMe](https://github.com/demisto/content/pull/18609/commits/b092fa4c28c53c8853c68d9e556d78363b393e79) even though CircleCI says they are not, and the ReadMe was generated using `update-release-notes`, so I'm not sure why they are not found
> 
> https://app.circleci.com/pipelines/github/demisto/content/162302/workflows/50568f28-0151-4b88-9f61-039fb4ea2c50/jobs/515945?invite=true#step-108-80 - Not sure what issue is?
> 
> https://app.circleci.com/pipelines/github/demisto/content/162302/workflows/50568f28-0151-4b88-9f61-039fb4ea2c50/jobs/515945?invite=true#step-108-94, https://app.circleci.com/pipelines/github/demisto/content/162302/workflows/50568f28-0151-4b88-9f61-039fb4ea2c50/jobs/515945?invite=true#step-108-102, https://app.circleci.com/pipelines/github/demisto/content/162302/workflows/50568f28-0151-4b88-9f61-039fb4ea2c50/jobs/515945?invite=true#step-108-114 --> DeleteContext. These are all called in sub-playbooks, and `auto` is set to True, so this is deleting the sub-playbook context, which is the goal (e.g if this was called in a loop). So this behavior is by design, not an error.
> 
> How do I move ahead with resolving these?

Hi :), I've send you through slack instruction on how to add the missing commands to the README, please look into it. In regard to the IN140 error code, it is a known issue and we’ll merge with it.

idovandijk reviewed (changes requested):
Thank you for this contribution. I will write points about the playbooks below:

### Sentinel One - Endpoint data collection
1. This check is insufficient because it checks whether an instance of the integration exists, but does not verify that it is enabled:
![image](https://user-images.githubusercontent.com/43602124/168619015-8cd81582-7f39-4092-b653-a80834424b51.png)

Please use the `IsIntegrationAvailable` script instead.

2. Does `so-get-agent-processes` produce outputs? If so, please declare them. If not - why?

3. Please add outputs to the playbook. It should output the important data that the playbook is used to retrieve - AKA the agent information and the process list.

4. Note that you named the playbook Sentinel One as opposed to SentinelOne in the other playbooks.


### SentinelOne - Check for New File Upload Activity

5. If validation fails because of DeleteContext with `all=yes`, you can specify the `Counter` key instead which would be a cleaner way to accomplish this.

6. Not sure I understand the section header about sleeps, MaxCounter and Timeout.

### SentinelOne - Check for New File Upload Activity - Inner Loop

7. Are you sure you need to delete context internally? I think that each loop should run with a new context for that subplaybook.

8. Can you set an initial default value for the MaxCounter?

9. Please define the sub-values for `SentinelOne.Activity` as outputs too, to make them visible from outside the playbook (they will be outputted anyway if you output `SentinelOne.Activity`, but will be harder to know what other keys are there).

10. We usually refrain from using `Sleep` in playbooks, but I read your explanation as to why you chose loop with sleeping. @David-BMS do you know whether GenericPolling can be used to check whether "no context was outputted at all" VS "context was outputted" rather than checking if a job is still running or has finished? Also @MLainer1 can you check whether they can use the code-based Generic Polling (in the integration instead of in the playbooks)? Supported only from XSOAR 6.2 though IIRC.

### SentinelOne - Fetch File from Endpoint

11. Why do we need to delete context in this playbook?

12. `Is SentinelOne v2 enabled?` same as point 1

13. Is `PossiblyInfected)(*&7890` the default password always?

14. Optional - consider adding this as a subplaybook to the generic `Get File Sample From Path - Generic V3` playbook. It would be a good addition and will improve your content exposure as well.

15.  You are using the Demisto Lock integration. Please make sure to check if it's enabled before proceeding in the playbook.

16. I'm a bit confused on how this actually works despite the note in the playbook description. What is the relation between an Activity and a file request? What if someone is creating an activity via the UI - will that interfere with the file request from XSOAR? It would be best if I can get a short walkthrough on this playbook. 

I appreciate your hard work on this content. I'd also be glad if we can schedule a demo to see an actual incident where those playbooks run, and if we can also look together at the `SentinelOne - Fetch File From Endpoint` playbook to see if it can be simplified or improved and to get a general better understanding of it.
Thanks again!

CraigWampler commented:
> 6\. about sleeps, MaxCounter

Thanks - got the Readme issues fixed. Next addressing Playbook issues

David-BMS commented:
@idovandijk Field polling is more suitable for what you described 
https://xsoar.pan.dev/docs/reference/playbooks/field-polling---generic

But if possible its best to update the integration command to have scheduled commands as seen here
https://xsoar.pan.dev/docs/integrations/scheduled-commands

CraigWampler commented:
> Thank you for this contribution. I will write points about the playbooks below:
> 
> ### Sentinel One - Endpoint data collection
> 1. This check is insufficient because it checks whether an instance of the integration exists, but does not verify that it is enabled:
>    ![image](https://user-images.githubusercontent.com/43602124/168619015-8cd81582-7f39-4092-b653-a80834424b51.png)
> 
> Please use the `IsIntegrationAvailable` script instead.

I did not write/modify this playbook 

> 2. Does `so-get-agent-processes` produce outputs? If so, please declare them. If not - why?

I did not write/modify this playbook 

> 3. Please add outputs to the playbook. It should output the important data that the playbook is used to retrieve - AKA the agent information and the process list.

I did not write/modify this playbook 

> 4. Note that you named the playbook Sentinel One as opposed to SentinelOne in the other playbooks.

I did not write/modify this playbook. In my opinion this one should be renamed to use "SentinelOne" 

> ### SentinelOne - Check for New File Upload Activity
> 5. If validation fails because of DeleteContext with `all=yes`, you can specify the `Counter` key instead which would be a cleaner way to accomplish this.

Made this change

> 6. Not sure I understand the section header about sleeps, MaxCounter and Timeout.

I am trying to explain what is not intuitive to me, which is that the MaxCounter paramter in block #1 is 4 times the input parameter.  Where/how would you put some text to explain that to a future maintainer of the code?

> ### SentinelOne - Check for New File Upload Activity - Inner Loop
> 7. Are you sure you need to delete context internally? I think that each loop should run with a new context for that subplaybook.

Modified to just delete SentinelOne.Activity

> 8. Can you set an initial default value for the MaxCounter?

Yes, now set

> 9. Please define the sub-values for `SentinelOne.Activity` as outputs too, to make them visible from outside the playbook (they will be outputted anyway if you output `SentinelOne.Activity`, but will be harder to know what other keys are there).


> 10. We usually refrain from using `Sleep` in playbooks, but I read your explanation as to why you chose loop with sleeping. @David-BMS do you know whether GenericPolling can be used to check whether "no context was outputted at all" VS "context was outputted" rather than checking if a job is still running or has finished? Also @MLainer1 can you check whether they can use the code-based Generic Polling (in the integration instead of in the playbooks)? Supported only from XSOAR 6.2 though IIRC.
> 
> ### SentinelOne - Fetch File from Endpoint
> 11. Why do we need to delete context in this playbook?

Why wouldn't we want to delete the context? We always want this PB (which runs as a sub-PB) to have an empty context. We don't ever want it to run with any data from a previous execution. It only deletes the Sub-PB context, so it seems to only be benefecial to me to have it delete the context at the beginning of each run. But I would love to understand the other side of that as to why we would not want to delete the context.

> 12. `Is SentinelOne v2 enabled?` same as point 1

Fixed

> 13. Is `PossiblyInfected)(*&7890` the default password always?

SentinelOne requires a password to protect the file with, and requires the password to have special characters, numbers, upper and lower case letters. I picked this as a reasonable default.

> 14. Optional - consider adding this as a subplaybook to the generic `Get File Sample From Path - Generic V3` playbook. It would be a good addition and will improve your content exposure as well.

Not planning to do this now.

> 15. You are using the Demisto Lock integration. Please make sure to check if it's enabled before proceeding in the playbook.

Added

> 16. I'm a bit confused on how this actually works despite the note in the playbook description. What is the relation between an Activity and a file request? What if someone is creating an activity via the UI - will that interfere with the file request from XSOAR? It would be best if I can get a short walkthrough on this playbook.

The way SentinelOne's API works is that you submit a file request, and whenever the agent is done retrieving a file, it posts it to the activity stream. So there is always a risk that a human (or another system using the API) has also asked for a file, and when the file requested by the human appears in the activity stream, our playbook grabs that one isntead. 

That said, I do not know of any way around this. By using Demisto Lock, we remove any chance of XSOAR (in another playbook) causing this condition. By using the StartTimestamp, we also reduce the risk of just getting the most recent file in the activity stream, and that being one that was requested by another process/human. We also filter our activity stream feed by the agent ID, so it has to be another user/process trying to get that a file from that host at the exact same time.

In case it's not clear from earlier discussion, I do not work for SentinelOne. My only goal in including these playbooks was to illustrate how to use the actions that I added to the integration. So I unfortunately do not have the ability to work with SentinelOne to request that they change their API to reduce this risk of a race condition. 

> 
> I appreciate your hard work on this content. I'd also be glad if we can schedule a demo to see an actual incident where those playbooks run, and if we can also look together at the `SentinelOne - Fetch File From Endpoint` playbook to see if it can be simplified or improved and to get a general better understanding of it. Thanks again!

CraigWampler commented:
> @idovandijk Field polling is more suitable for what you described https://xsoar.pan.dev/docs/reference/playbooks/field-polling---generic
> 
> But if possible its best to update the integration command to have scheduled commands as seen here https://xsoar.pan.dev/docs/integrations/scheduled-commands

My main goal with this was to contribute the changes to the integration back. The playbooks were meant to illustrate how to use the new actions that I added. If these changes are necessary in order for the playbooks to be accepted, would it be possible to just remove the playbooks from the submission for now?

idovandijk commented:
@CraigWampler while I understand your view, the playbooks still have to be reviewed. I will point out only the things that must be done in order to release them. I'll refer to your answers by the numbers:

1. If we'd like to keep this playbook, it needs to at least check if the integration is enabled properly and to have the outputs of the data it was meant to retrieve (E.g., - agent info and processes). If not, please remove the playbook from this PR.
6. I understand it now. This game of maxcounter + hardcoded 15 seconds for sleep is confusing. Also it would not be accurate, because other tasks in the playbook run as well, which take time, so if I choose a MaxCounter of 4 then it could still take longer technically. We have the built-in sleep mechanism as part of Loops: 
![image](https://user-images.githubusercontent.com/43602124/169011800-bdfa5998-f067-40e7-97c2-6ccb5c3c55c7.png)
11. There is no disadvantage to deleting context, but I do not see where this playbook is being looped (where is this playbook called from?)
13. That's fine, but if password is mandatory let's also mark this as mandatory in the inputs.
14. That is also fine.
16. This is the main issue in this PR, and while your demonstrated solution is optimal it's not ideal, and it's still prone to race conditions and interference from the UI. SentinelOne needs to add a correlation between the file request to the file entry in the activity feed. If this is not something that is going to be added with a proper polling then please remove the playbook.

Feel free to remove playbooks that you do not intend to address, but please remember to also remove their corresponding READMEs and to remove their part in the release notes. Also, note that some playbooks were not formatted using `demisto-sdk format`, please format them.

Lastly I would still need to see an execution of those either in a demo or a video demonstrating usage, as we do not have a way to simulate this ourselves.
Thanks again.

CraigWampler commented:
I just pushed a new commit that removes the playbooks to make this submission simpler. I also left a note in Slack about some times to perform a demo.

idovandijk commented:
> I just pushed a new commit that removes the playbooks to make this submission simpler. I also left a note in Slack about some times to perform a demo.

Cool, in the demo we can simply observe the updated PR in action, for me it would be sufficient if you have an incident that ran with the playbooks that you are contributiing. Messaged back in DFIR
